components/eamxx/src/physics/spa/spa_functions_impl.hpp

```diff
@@ -425,12 +444,13 @@ perform_vertical_interpolation(
  * pressure profile that is constructed in spa_main is
  * a) the current length, and
  * b) ensures that whatever the target pressure profile is, it's within the
- *    the range of the source data.                                        
+ *    the range of the source data.
  */
 template<typename S, typename D>
+template<typename DataReader>
 void SPAFunctions<S,D>
 ::update_spa_data_from_file(
-          AtmosphereInput&  spa_data_reader,
+          DataReader&       spa_data_reader,
     const int               time_index, // zero-based
           AbstractRemapper& spa_horiz_interp,
           SPAInput&         spa_input)
```

Comment:
Abstract the SPA update functions (this one and the one below) on `DataReader` type, where the `spa_data_reader` object only needs to define a function `read_variables(time_index)` that reads file data to fields defined on the remapper source grid. Standard grids use `AtmosphereInput` and can have a remapper, IOP uses `IOPReader` and is required to have identity remapper.

Reply:
When I wrote the interface, I used a reference (and deref-ed the pointer at call site) for the remapper, but you could easily do
```
update_spa_data_from_file(const std::shared_ptr<AtmosphereInput>& scorpio_reader,
						 const std::shared_ptr<IOPReader>& iop_reader,
						 ...)
{
  if (iop_reader) {
    iop_reader->blah(...);
  } else {
    scorpio_reader->blah2(...);
  }
}
```
so that the two readers don't necessarily need to have the same interface.

Up to you. If you like the template version better, that's fine with me. Just wanted to point out you can change the interface, if you think it's better.

Reply:
I'll make this change so that I can pass in a timestamp for the IOP reader.